File: frontend/src/types/keci/userSeriesAccess.ts
import type { User } from './user'
import type { PodcastSeries } from './podcast'
import type { Article } from './content'

export interface UserSeriesAccess {
  userSeriesAccessId: number
  userId: number
  seriesId?: number | null
  articleId?: number | null
  currentAccessibleSequence: number
  updatedAt: string
  user?: User
  podcastSeries?: PodcastSeries
  article?: Article
}

export interface CreateUserSeriesAccess {
  userId: number
  seriesId?: number | null
  articleId?: number | null
  currentAccessibleSequence: number
}

export interface UpdateUserSeriesAccess {
  currentAccessibleSequence: number
}

export interface GrantAccessRequest {
  userId: number
  seriesId: number
  articleId?: number | null
}

export interface RevokeAccessRequest {
  userId: number
  seriesId?: number | null
  articleId?: number | null
}

export interface UserSeriesAccessFilter {
  search?: string
  userId?: number
  seriesId?: number
  hasAccess?: boolean
}

export interface UserSeriesAccessFormData {
  userId: number
  seriesId: number
  currentAccessibleSequence: number
  nextUnlockDate: string
  currentPositionInSeconds: number
}

File: KeciApp.API/Models/UserSeriesAccess.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace KeciApp.API.Models;

public class UserSeriesAccess
{
    [Key] public int UserSeriesAccessId { get; set; }

    [Required] [ForeignKey("User")] public int UserId { get; set; }

    [ForeignKey("Article")] public int? ArticleId { get; set; }

    [ForeignKey("PodcastSeries")]
    public int? SeriesId { get; set; }

    [Required] public int CurrentAccessibleSequence { get; set; }

    [Required] public DateTime UpdatedAt { get; set; }

    // Navigation properties
    public User User { get; set; }
    public PodcastSeries PodcastSeries { get; set; }
    public Article? Article { get; set; }
}